data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java

```diff
@@ -565,15 +562,40 @@ public void requireStoreAvailable() {
 
   private List<ValidatorResponse> getValidators(
       final List<Integer> validatorIndices,
+      final List<ValidatorStatus> statusFilter,
       final tech.pegasys.teku.datastructures.state.BeaconState state) {
-    return validatorIndices.stream()
-        .map(index -> ValidatorResponse.fromState(state, index))
-        .collect(toList());
+    final List<ValidatorResponse> result;
+
+    if (validatorIndices.isEmpty()) {
+      final int validatorCount = state.getValidators().size();
+      result = new ArrayList<>();
+      for (int i = 0; i < validatorCount; i++) {
+        result.add(ValidatorResponse.fromState(state, i));
+      }
+    } else {
+      result =
+          validatorIndices.stream()
+              .map(index -> ValidatorResponse.fromState(state, index))
+              .collect(toList());
+    }
+
+    if (statusFilter.isEmpty()) {
+      return result;
+    }
+    return result.stream().filter(v -> statusFilter.contains(v.status)).collect(toList());
```

Comment:
There's  a weird mix of iteration and streaming here that's winding up converting to list multiple times.  It seems like this would be a lot cleaner if we passed in functions to select and filter the validators instead of lists.  I've done a rough spike of this (commenting out a bunch of tests to avoid flowing the changes round everywhere): https://github.com/ConsenSys/teku/commit/efd37ca00cc39b8f137b1b76ea8c94c0898086ca

The key thing is it introduces a `ValidatorSelector` interface which is essentially a function that takes a `BeaconState` and returns an IntStream of validator indices to use from that state.  So when no validator IDs are provided, the validator selector is `state -> IntStream.range(0, state.getValidators.size())` - ie all validators.  Otherwise it just maps the list of validator selectors to an index.  An `IntStream` is returned rather than `List<Integer>` so each entry is processed in turn without having to load them all in memory (100k Integer instances isn't particularly friendly to the GC).